File: clojure/anagram/src/anagram.clj
(ns anagram
  (:require [clojure.string :as str]))

(defn- anagram? [s1 s2]
  (let [lc1 (str/lower-case s1)
        lc2 (str/lower-case s2)]
    (and (= (sort lc1) (sort lc2))
         (not= lc1 lc2))))

(defn anagrams-for [s choices]
  (filter #(anagram? s %) choices))

File: elixir/anagram/anagram.exs
defmodule Anagram do
  @doc """
  Returns all candidates that are anagrams of, but not equal to, 'base'.
  """
  @spec match(String.t, [String.t]) :: [String.t]
  def match(base, candidates) do
    candidates
    |> Enum.filter(&(anagram?(&1, base)))
  end

  defp anagram?(s1, s2) do
    ds1 = String.downcase(s1)
    ds2 = String.downcase(s2)
    sorted_chars(ds1) == sorted_chars(ds2) && ds1 != ds2
  end

  defp sorted_chars(s) do
    s
    |> String.graphemes
    |> Enum.sort
  end
end
